asserts/ifacedecls.go

```diff
@@ -0,0 +1,187 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2015 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package asserts
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+)
+
+type attrMatcher interface {
+	match(interface{}) error
+}
+
+// compileAttrMatcher compiles an attrMatcher derived from constraints,
+// context can be "top" or "map", alternatives have no influence on it
+func compileAttrMatcher(context string, constraints interface{}) (attrMatcher, error) {
+	switch x := constraints.(type) {
+	case map[string]interface{}:
+		return compileMapAttrMatcher(x)
+	case []interface{}:
+		// TODO: disallow nesting alternatives directly inside alernatives?
+		return compileAltAttrMatcher(context, x)
+	case string:
+		if context == "top" {
```

Comment:
Should we use a dotted path as the context (thus empty here) so that we can provide neat errors? For regular expression compilation errors, etc.